### ISSUE TITLE
Allow inverting series highlight behavior: change only color/alpha of the hovered series

### ISSUE THREAD

**Issue by juliusv** [leeoniya/uPlot#990]
When focusing a series by hovering, the `focus.alpha` setting allows you to de-emphasize all other series by adding transparency to them. I find that effect a bit too visually jarring and would like to achieve the opposite, where only the series that is currently focused gets emphasized somehow, while all others keep their style. Ideally the emphasis wouldn't be alpha-based (since alpha is already 1), but would allow me to change the color or width of the series.

**Comment by juliusv** [comment]
The one way I got this working with the current uPlot is by remembering the selected series index with a setSeries hook like this:

```typescript
  let selectedSeriesIdx: number | null = null;

  // ...

      setSeries: [
        (u: uPlot, seriesIdx: number | null, _opts: Series) => {
          if (seriesIdx === selectedSeriesIdx) {
            return;
          }

          if (selectedSeriesIdx !== null) {
            u.series[selectedSeriesIdx].width = 1.5;
          }
          if (seriesIdx !== null) {
            u.series[seriesIdx].width = 2.2;
          }
          selectedSeriesIdx = seriesIdx;
          u.redraw();
        },
      ],
```

However, the required full redraw makes hovering sluggish when there are many series, so I probably won't be able to use this solution.

**Comment by leeoniya** [comment]
yeah, this is the way that https://leeoniya.github.io/uPlot/demos/focus-cursor.html is implemented.

there's no getting away from full redraw. but the paths are not re-generated, so it shouldnt be horribly expensive just to clear the canvas and re-paint the cached Path2Ds.

do you have a demo or jsFiddle that shows the slowness?

related: https://github.com/leeoniya/uPlot/issues/787

that being said, it probably makes sense to have an api for `cursor.focus` that does not require the user externally remembering the original line width / style / stroke / fill.

**Comment by leeoniya** [comment]
> that being said, it probably makes sense to have an api for `cursor.focus` that does not require the user externally remembering the original line width / style / stroke / fill.

so, i started down this path in https://github.com/leeoniya/uPlot/tree/focus-styles, and it ends up being more complex than i think is worth it.

mutating the series widths works well, and you don't have to call redraw again.

see how it's done here:

**Comment by leeoniya** [comment]
ok, i tried it out and yeah, if the `focus.alpha: 1` it will bail out of redraw. so you do need the manual `u.redraw(false)` call. the `false` argument will prevent it from regenerating the series paths, so is very fast. i added a demo in https://github.com/leeoniya/uPlot/commit/d1f7abb9984f36c6d4b5a98ed6514bf36c873a61.

https://github.com/leeoniya/uPlot/blob/d1f7abb9984f36c6d4b5a98ed6514bf36c873a61/demos/focus-cursor.html#L168-L179

you'll notice that the mutation strategy leaves the series points and hover points (in the dom) not changing size. also the hover point does not change color until you move to another x index. mutating `series.points.width` also isnt so straightforward.

maybe in the future we can make use of Canvas shadow properties:

```
ctx.shadowColor = "red";
ctx.shadowOffsetX = 5;
ctx.shadowOffsetY = 5;
ctx.shadowBlur = 1;
```

i'm afraid there's not a lot of straightforward solutions here. as much as i want to help, this is the first time in 5 years this request has come up, and the juice just isn't worth the squeeze.

**Comment by leeoniya** [comment]
thinking about it _even more_, it's not enough to change style/thickness/color. you would _also_ have to change its rendering order, otherwise it would be occluded 100 other line series.
